src/numerical_method.py

```diff
@@ -184,4 +191,66 @@ def __init__(self, exp: float, strike: float, nr_steps: int = 1, up_step_mult: f
         self.nr_steps = nr_steps
         self.up_step_mult = up_step_mult
         self.down_step_mult = down_step_mult
-        self.strike = strike
\ No newline at end of file
+        self.strike = strike
+
+
+class BlackScholesPDE(PDEMethod):
+    def __init__(self, model: MarketModel, params: PDEParams):
+        super().__init__(model, params)
+        self.exp = params.exp
+        self.strike = params.strike
+        self.sigma = model.get_vol(self.exp, model.get_initial_spot()/self.strike)
+        self.t_step = params.t_step
+        self.und_step = params.s_step
+        self._derivative_type = params.contract_type
+        self.S_min = params.S_min
+        self.S_max = params.S_max
+        self.ns_steps = int(np.round((self.S_max - self.S_min) / float(self.und_step)))  # Number of stock price steps
+        self.nt_steps = int(np.round(params.exp / float(self.t_step)))   # Number of time steps
+        self._interest_rate = model.get_rate()
+        self.grid = np.zeros((self.nt_steps + 1, self.ns_steps + 1))
+        self.P, self.Q, self.R = self.tridiagonal_matrix()
+
+    def setup_boundary_conditions(self):
+        if self._derivative_type == PutCallFwd.CALL:
+            self.grid[0, :] = np.maximum(np.linspace(self.S_min, self.S_max, self.ns_steps + 1) - self.strike, 0)
```

Comment:
Could you please explore if we can use the contract's payoff method here? You can ask for the contract object as input. The vectorized form might be replaced with a loop, but this way we can incorporate other contracts as well. E.g. it would be nice if it worked for a Forward as well.